File: scripts/tank/report.py
"""The one finding shape every asset-door stage emits, and its deterministic renderings.

Text and JSON are renderings of this shape; the exit status is `exit_code`. A check's severity is
compiled into its `Check` beside its id — there is no severity file, allowlist, per-asset override
or warning budget anywhere in the door.

The Python half of the door spans two interpreters: the Blender source pass runs inside Blender's
embedded CPython, the derivation verifier and the outer wrapper run under the system one. So this
module is stdlib-only and never imports `bpy`.

`src/bake/report.rs` is the same shape in Rust, and the two agree on stage and severity ordering:
one report reads the same rows in the same order whichever half produced them.
"""

from __future__ import annotations

import enum
import json
from dataclasses import dataclass
from typing import Optional, Sequence, Tuple


class Stage(enum.IntEnum):
    """Which pass a check belongs to. The report's primary sort key, declared in pipeline order.

    `DOOR` precedes every pass: it carries the door's own mechanical refusals, which happen before
    a check of the asset can be reached. The Rust half never emits it and so does not declare it.
    """

    DOOR = 0
    SOURCE = 1
    CONSUMER = 2
    DERIVATION = 3


class Severity(enum.IntEnum):
    """A finding's tier. Declared in report order, so a sort puts the build-failing rows first."""

    ERROR = 0
    WARNING = 1
    INFO = 2

    @property
    def label(self) -> str:
        return _SEVERITY_LABELS[self]


_SEVERITY_LABELS = {
    Severity.ERROR: "error",
    Severity.WARNING: "warning",
    Severity.INFO: "info",
}

_SEVERITY_PLURALS = {
    Severity.ERROR: "errors",
    Severity.WARNING: "warnings",
    Severity.INFO: "info",
}


class SubjectKind(enum.IntEnum):
    """What a finding is about. Declared outermost-first, which is also the report's sort order."""

    DOOR = 0
    FILE = 1
    SCENE = 2
    OBJECT = 3
    MESH = 4
    MATERIAL = 5


_SUBJECT_LABELS = {
    SubjectKind.DOOR: "door",
    SubjectKind.FILE: "file",
    SubjectKind.SCENE: "scene",
    SubjectKind.OBJECT: "object",
    SubjectKind.MESH: "mesh",
    SubjectKind.MATERIAL: "material",
}


@dataclass(frozen=True)
class Subject:
    """The thing a finding names: kind, name, and the element inside it — a modifier, a shape-key
    datablock, a transform channel — where the check found something smaller than the whole
    subject."""

    kind: SubjectKind
    name: str
    element: Optional[str] = None

    def sort_key(self) -> Tuple[int, str, str]:
        return (int(self.kind), self.name, self.element or "")

    def __str__(self) -> str:
        text = "{} `{}`".format(_SUBJECT_LABELS[self.kind], self.name)
        return "{} {}".format(text, self.element) if self.element else text


@dataclass(frozen=True)
class Check:
    """One check: its id, the pass and tier compiled in beside it, and the law it enforces."""

    id: str
    stage: Stage
    severity: Severity
    #: The law, stated as the condition that must hold.
    law: str


@dataclass(frozen=True)
class Finding:
    """One violation: which check, of what, with what was measured and what repairs it."""

    check: Check
    subject: Subject
    #: What was measured — the numbers, names and coordinates the verdict rests on.
    evidence: str
    #: The concrete edit that makes the law hold.
    repair: str

    def sort_key(self):
        """Stage, severity, check id, subject, element — the declared report order. Evidence breaks
        a remaining tie, so two findings that differ at all still order deterministically."""
        return (
            int(self.check.stage),
            int(self.check.severity),
            self.check.id,
            self.subject.sort_key(),
            self.evidence,
        )

    def as_dict(self) -> dict:
        return {
            "check": self.check.id,
            "stage": self.check.stage.name.lower(),
            "severity": self.check.severity.label,
            "subject": {
                "kind": _SUBJECT_LABELS[self.subject.kind],
                "name": self.subject.name,
                "element": self.subject.element,
            },
            "evidence": self.evidence,
            "law": self.check.law,
            "repair": self.repair,
        }

    def __str__(self) -> str:
        return (
            "{id} {severity}: {subject}\n"
            "  measured: {evidence}\n"
            "  law:      {law}\n"
            "  repair:   {repair}"
        ).format(
            id=self.check.id,
            severity=self.check.severity.label,
            subject=self.subject,
            evidence=self.evidence,
            law=self.check.law,
            repair=self.repair,
        )


def sorted_findings(findings: Sequence[Finding]) -> list:
    """Put a report in its declared order. Every producer sorts before it returns, so the console,
    the JSON and a GUI popup read the same rows in the same order."""
    return sorted(findings, key=Finding.sort_key)


def has_error(findings: Sequence[Finding]) -> bool:
    """Whether the report fails the build."""
    return any(finding.check.severity == Severity.ERROR for finding in findings)


def exit_code(findings: Sequence[Finding]) -> int:
    """Non-zero exactly when the report holds an error. Warnings never change exit status."""
    return 1 if has_error(findings) else 0


def counts(findings: Sequence[Finding]) -> dict:
    """How many findings of each tier, every tier present as a key."""
    tally = {severity: 0 for severity in Severity}
    for finding in findings:
        tally[finding.check.severity] += 1
    return tally


def summary(findings: Sequence[Finding]) -> str:
    """The one-line tally a GUI popup can show whole. It never replaces the rows: the complete
    report is the console and the JSON."""
    tally = counts(findings)
    return ", ".join(
        "{} {}".format(
            tally[severity],
            severity.label if tally[severity] == 1 else _SEVERITY_PLURALS[severity],
        )
        for severity in Severity
    )


def render_text(findings: Sequence[Finding]) -> str:
    """The complete text rendering — every finding, one block each."""
    return "".join("{}\n".format(finding) for finding in findings)


def render_json(findings: Sequence[Finding]) -> str:
    """The complete JSON rendering — the same rows in the same order, one object each."""
    return json.dumps({"findings": [finding.as_dict() for finding in findings]}, indent=2)

File: src/bake/report.rs
//! The one finding shape every asset-door stage emits, and its deterministic rendering.
//!
//! Text, JSON and GUI output are renderings of this shape; the exit status is [`has_error`]. A
//! check's severity is compiled into its [`Check`] definition beside its id — there is no severity
//! file, allowlist, per-asset override or warning budget anywhere in the door.

use std::fmt;

/// Which pass a check belongs to. The report's primary sort key, declared in pipeline order.
///
/// `Door` precedes every pass: it carries the door's own mechanical refusals, which happen before a
/// check OF the asset can be reached — an input the door failed to supply is not a defect of the
/// model.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum Stage {
    /// The wrapper's own mechanical refusals (`door.*`).
    Door,
    /// The `.blend` source pass (`L1.*`).
    Source,
    /// The shared consumer contract (`L2.*`) — this module's callers.
    Consumer,
    /// The GLB derivation: texture encode and repack (`D.*`).
    Derivation,
}

/// A finding's tier. Declared in report order, so a sort puts the build-failing rows first.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum Severity {
    /// A real refusal: the asset does not enter the sim. Always build-failing.
    Error,
    /// Generic evidence needing human judgement. Never changes exit status.
    Warning,
    /// Census and change evidence only.
    Info,
}

impl Severity {
    fn label(self) -> &'static str {
        match self {
            Self::Error => "error",
            Self::Warning => "warning",
            Self::Info => "info",
        }
    }
}

/// What a finding is about.
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum SubjectKind {
    /// The door itself; the subject name is the stage that refused.
    Door,
    /// The `.tank.ron` spec sheet; the subject name is its path.
    Spec,
    /// The GLB document as a whole; the subject name is its path.
    Document,
    /// One glTF node; the subject name is the node name.
    Node,
}

impl SubjectKind {
    fn label(self) -> &'static str {
        match self {
            Self::Door => "door",
            Self::Spec => "spec",
            Self::Document => "document",
            Self::Node => "node",
        }
    }
}

/// The thing a finding names: kind, name, and the element inside it (a primitive, a triangle, an
/// edge, a shell) where the check found something smaller than the whole subject.
#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub struct Subject {
    pub kind: SubjectKind,
    pub name: String,
    pub element: Option<String>,
}

impl Subject {
    pub(crate) fn door(stage: &str) -> Self {
        Self {
            kind: SubjectKind::Door,
            name: stage.to_owned(),
            element: None,
        }
    }

    pub(crate) fn spec(path: &std::path::Path) -> Self {
        Self {
            kind: SubjectKind::Spec,
            name: path.display().to_string(),
            element: None,
        }
    }

    pub(crate) fn document(path: &std::path::Path) -> Self {
        Self {
            kind: SubjectKind::Document,
            name: path.display().to_string(),
            element: None,
        }
    }

    pub(crate) fn node(name: &str) -> Self {
        Self {
            kind: SubjectKind::Node,
            name: name.to_owned(),
            element: None,
        }
    }

    pub(crate) fn element(mut self, element: impl Into<String>) -> Self {
        self.element = Some(element.into());
        self
    }
}

impl fmt::Display for Subject {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} `{}`", self.kind.label(), self.name)?;
        match &self.element {
            Some(element) => write!(f, " {element}"),
            None => Ok(()),
        }
    }
}

/// One check: its id, the pass and tier compiled in beside it, and the law it enforces.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct Check {
    pub id: &'static str,
    pub stage: Stage,
    pub severity: Severity,
    /// The law, stated as the condition that must hold.
    pub law: &'static str,
}

/// One violation: which check, of what, with what was measured and what repairs it.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct Finding {
    pub check: &'static Check,
    pub subject: Subject,
    /// What was measured — the numbers, names and coordinates the verdict rests on.
    pub evidence: String,
    /// The concrete edit that makes the law hold.
    pub repair: String,
}

impl Finding {
    pub(crate) fn new(
        check: &'static Check,
        subject: Subject,
        evidence: impl Into<String>,
        repair: impl Into<String>,
    ) -> Self {
        Self {
            check,
            subject,
            evidence: evidence.into(),
            repair: repair.into(),
        }
    }

    /// Stage, severity, check id, subject, element — the declared report order. Evidence breaks a
    /// remaining tie, so two findings that differ at all still order deterministically.
    fn sort_key(&self) -> (Stage, Severity, &'static str, &Subject, &str) {
        (
            self.check.stage,
            self.check.severity,
            self.check.id,
            &self.subject,
            self.evidence.as_str(),
        )
    }
}

impl fmt::Display for Finding {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(
            f,
            "{} {}: {}",
            self.check.id,
            self.check.severity.label(),
            self.subject
        )?;
        writeln!(f, "  measured: {}", self.evidence)?;
        writeln!(f, "  law:      {}", self.check.law)?;
        write!(f, "  repair:   {}", self.repair)
    }
}

/// Put a report in its declared order. Every producer sorts before it returns, so the console, a
/// panic message and a future JSON rendering read the same rows in the same order.
pub(crate) fn sorted(mut findings: Vec<Finding>) -> Vec<Finding> {
    findings.sort_by(|a, b| a.sort_key().cmp(&b.sort_key()));
    findings
}

/// Whether the report fails the build.
pub fn has_error(findings: &[Finding]) -> bool {
    findings
        .iter()
        .any(|finding| finding.check.severity == Severity::Error)
}

/// The complete text rendering — every finding, one block each.
pub fn render(findings: &[Finding]) -> String {
    findings
        .iter()
        .map(|finding| format!("{finding}\n"))
        .collect()
}
